pkg/danm/danm.go

```diff
@@ -183,13 +191,115 @@ func extractConnections(args *cniArgs) error {
+}
+
+func getRegisteredDevices(args *cniArgs, cniType string)([]string,error){
+  resourceMap := make(map[string]*checkpoint_types.ResourceInfo)
+  if string(args.podUid) != "" {
+    checkpoint, err := checkpoint_utils.GetCheckpoint()
+    if err != nil {
+      return nil, errors.New("failed to instantiate checkpoint object due to:" + err.Error())
+    }
+    resourceMap, err = checkpoint.GetComputeDeviceMap(string(args.podUid))
+    if err != nil {
+      return nil, errors.New("failed to retrieve Pod info from checkpoint object due to:" + err.Error())
+    }
+  }
+  prefix, err := getResourcePrefix(args, cniType)
+  if err != nil {
+    return nil, errors.New("failed to get resource prefix from config map due to:" + err.Error())
+  }
+  var registeredDevices []string
+  for resourcename, resources := range resourceMap {
+    if strings.HasPrefix(resourcename,prefix) {
+      registeredDevices = append(registeredDevices, resources.DeviceIDs...)
+    }
+  }
+  return registeredDevices, nil
+}
+
+func getSriovInterfaces(args *cniArgs)(map[string]int,map[string]string,error){
+  danmClient, err := createDanmClient(args.stdIn)
+  if err != nil {
+    return nil, nil, errors.New("failed to create DanmClient due to:" + err.Error())
+  }
+  sriovInterfaces := make(map[string]int)
+  interfaceDeviceMap := make(map[string]string)
+  for _, interfac := range args.interfaces {
+    danmnet, err := danmClient.DanmV1().DanmNets(args.nameSpace).Get(interfac.Network, meta_v1.GetOptions{})
+    if err != nil || danmnet.ObjectMeta.Name == ""{
+      return nil, nil, errors.New("NID:" + interfac.Network + " in namespace:" + args.nameSpace + " cannot be GET from K8s API server!")
+    }
+    if danmnet.Spec.NetworkType == "sriov" {
+      sriovInterfaces[interfac.Network]++
+      interfaceDeviceMap[interfac.Network] = danmnet.Spec.Options.Device
+    }
+  }
+  return sriovInterfaces, interfaceDeviceMap, nil
+}
+
+func validateSriovNetworkRequests(sriovInterfaces map[string]int, sriovDevices []string, interfaceDeviceMap map[string]string) error {
+  requiredVfonPf := make(map[string]int)
+  requestedVfonPf := make(map[string]int)
+  for interfac, count := range sriovInterfaces {
+    requiredVfonPf[interfaceDeviceMap[interfac]] = count
+  }
+  for _, device := range sriovDevices {
+    pf, err := sriov_utils.GetPfName(device)
+    if err != nil {
+      return errors.New("failed to get the name of the sriov PF for device "+ device +" due to:" + err.Error())
+    }
+    requestedVfonPf[pf]++
+  }
+  eq := reflect.DeepEqual(requiredVfonPf, requestedVfonPf)
+  if eq {
+    return nil
+  }
+  log.Printf("Required SR IOV resources: %v", requiredVfonPf)
+  log.Printf("Requested SR IOV resources: %v", requestedVfonPf)
+  return errors.New("requested and required sriov resources are not matching in the Pod definition")
+}
+
 func setupNetworking(args *cniArgs) (*current.Result, error) {
+  sriovInterfaces, interfaceDeviceMap, err := getSriovInterfaces(args)
```

Comment:
the code is not generic, and the logic is wrong. You unnecessarily increased REST traffic towards the API server by a big margin, which can lead to serious performance issues under heavy load

here is the right logic:
- you read the specific DanmNet for the specific interface inside the for loop. Once. Only once, and only the referenced. You don't read it anymore, anywhere else, e.g. inside cnidel. You re-use the same DanmNet for the rest of network provisioning
- you check if its NetworkType has any Devices configured in the ConfigMap
- if not, you do everything as right now
- if yes, you read the allocated devices with the configured name from the checkPoint file, once. You store it in a hashmap or something, for the networkType
- you choose the Xth device from the device list for this specific network provisioning operation.
- you pass it down to the network provisioner, than let it do its job. you can add it to the Interfaces struct
- you increase X for the NetworkType. You maintain X in the same hashmap, where you store the DeviceIDs

Next iteration:
- oh look, anothert sriov type DanmNet
- oh look, I already read out the allocated Devices
- oh okay, so X is at one, therefore I need to configure the second Device now for this EndPoint, and increase X

Reply:
multiple DeviceNames matching for the prefix configured for one NetworkType needs to be reconsidered though
the logic applies whatever is the case: you only read one DanmNet once for one DanmEp, you simply pass down the info to the delegating code, and you maintain an intermediate data structure to track which devices were allocated already during this operation

Reply:
Thank you for the comment. The code has been refactored accordingly. The validation part is limited to comparing the VF's PF name to the DanmNet's host_device.
I tested use-cases (not all of them) and this solution is much more permissive with deviant configurations (wrong Resource Request, missing DP, etc.). Please comment on.

Reply:
even that validation is not needed, as it is SR-IOV specific thing that host_device, and DeviceName are related
we are getting there BTW, still some optimizations are needed (and the corrections to the smaller scope comments from the previous batch)